firmware/src/rtc.c

```diff
@@ -5,33 +5,94 @@
 
 ClockStatus Rtc_Init(Rtc *self)
 {
-    return CLOCK_OK;
+    if (self->getHour != NULL && self->getHour()) {
+        return CLOCK_OK;
+    }
+    return CLOCK_FAIL;
 }
 
 ClockStatus Rtc_SetTime(Rtc *self, RtcTime *time)
 {
-    if (self->setRtcTime != NULL)
-    {
-        self->setRtcTime(
+    if (self->setRtcTime == NULL) {
+        return CLOCK_FAIL;
+    }
+
+    self->setRtcTime(
             time->hr,
             time->min,
-            time->sec,
-            time->date,
-            time->month,
-            time->dow,
-            time->year);
+            time->sec);
+
+    if (self->setDay == NULL || self->setMonth == NULL) {
+        return CLOCK_FAIL;
     }
+
+    self->setDay(1);
+    self->setMonth(1);
+
     return CLOCK_OK;
 }
 
 ClockStatus Rtc_GetTime(Rtc *self, RtcTime *time)
 {
-    if (self->getDayOfWeek != NULL)     { time->dow = self->getDayOfWeek(); }
-    if (self->getDate != NULL)          { time->date = self->getDate(); }
-    if (self->getMonth != NULL)         { time->month = self->getMonth(); }
-    if (self->getYear != NULL)          { time->year = self->getYear(); }
-    if (self->getHour != NULL)          { time->hr = self->getHour(); }
-    if (self->getMinute != NULL)        { time->min = self->getMinute(); }
-    if (self->getSecond != NULL)        { time->sec = self->getSecond(); }
+    if (self->getHour == NULL
+        || self->getMinute == NULL
+        || self->getSecond == NULL) {
+        return CLOCK_FAIL;
+    }
+
+    time->hr = self->getHour();
+    time->min = self->getMinute();
+    time->sec = self->getSecond();
+
+    return CLOCK_OK;
+}
+
+ClockStatus Rtc_IsValid(Rtc *self)
+{
+    if (self->getDay == NULL || self->getMonth == NULL) {
+        return CLOCK_FAIL;
+    }
+
+    if (self->getDay() > MAX_NO_DAYS_RESET || self->getMonth() > 1) {
+        return CLOCK_FAIL;
+    }
+    return CLOCK_OK;
+}
+
+ClockStatus Rtc_SetAlarm(Rtc *self, RtcTime *time, ALARM_ID id)
+{
+    if ((id != ALARM && id != TIMER) || self->setAlarm == NULL) {
+        return CLOCK_FAIL;
+    }
+
+    self->setAlarm(id, time->hr, time->min, time->sec);
+
+    return CLOCK_OK;
+}
+
+ClockStatus Rtc_GetAlarm(Rtc *self, RtcTime *time, ALARM_ID id)
+{
+    if ((id != ALARM && id != TIMER)
+            || self->getAlarmHour == NULL
+            || self->getAlarmMinute == NULL
+            || self->getAlarmSecond == NULL) {
+        return CLOCK_FAIL;
+    }
+
+    time->hr = self->getAlarmHour(id);
+    time->min = self->getAlarmMinute(id);
+    time->sec = self->getAlarmSecond(id);
+
+    return CLOCK_OK;
+}
+
+ClockStatus Rtc_EnableAlarm(Rtc *self, ALARM_ID id, ALARM_STATUS enable)
+{
+    if ((id != ALARM && id != TIMER) || self->enableAlarm == NULL) {

```

Comment:
Technically don't need the id!=ALARM id!=TIMER check since only parameters of type ALARM_ID can be passed in. Up to you if you want to keep it

Reply:
You're write, the less code the better so I will take it out